main/core/src/eval/Evaluator.scala

```diff
@@ -362,15 +382,302 @@ case class Evaluator(home: os.Path,
       }
     }
 
+    logEnd.foreach(_())
+
     multiLogger.close()
 
     (newResults, newEvaluated)
   }
 
   def resolveLogger(logPath: Option[os.Path], logger: Logger): Logger = logPath match{
     case None => logger
-    case Some(path) => MultiLogger(logger.colored, log, FileLogger(logger.colored, path, debugEnabled = true))
+    case Some(path) => MultiLogger(logger.colored, log, new FileLogger(logger.colored, path, debugEnabled = true))
   }
+
+  type Terminal = Either[Task[_], Labelled[Any]]
+  type TerminalGroup = (Terminal, Agg[Task[_]])
+
+  def evaluateParallel(
+    goals: Agg[Task[_]],
+    threadCount: Int,
+    reporter: Int => Option[BuildProblemReporter] = (int: Int) => Option.empty[BuildProblemReporter],
+    testReporter: TestReporter = DummyTestReporter,
+    logger: Logger = log
+  ): Evaluator.Results = {
+    os.makeDir.all(outPath)
+
+    val startTime = System.currentTimeMillis()
+
+    // we need to collect all relevant logging info while developing
+    val evalLog = MultiLogger(true, this.log,
+      new FileLogger(false, outPath / "evaluator.log", true, append = true) {
+        override def debug(s: String) = super.debug(s"${System.currentTimeMillis() - startTime} [${Thread.currentThread().getName()}] ${s}")
+      })
+
+    evalLog.info(s"Using experimental parallel evaluator with ${threadCount} threads")
+    evalLog.debug(s"Start time: ${new java.util.Date()}")
+
+    val (sortedGroups, transitive) = Evaluator.plan(rootModule, goals)
+
+    // Mutable collector for all evaluated tasks
+    val evaluated = new Agg.Mutable[Task[_]]
+
+    // Mutable collector for all task results
+    @volatile var results = Map.empty[Task[_], Result[(Any, Int)]]
+
+    type Timing = (Either[Task[_], Labelled[_]], Int, Boolean)
+
+    case class FutureResult(task: TerminalGroup, time: Int, result: Evaluated)
+
+    // Mutable collector for timings
+    val timings = new mutable.ArrayBuffer[Timing](sortedGroups.keyCount)
+
+    // Increment the counter message by 1 to go from 1/10 to 10/10
+    val nextCounterMsg = new Evaluator.NextCounterMsg(sortedGroups.keyCount)
+
+    // TODO: check for interactivity
+    // TODO: make sure, multiple goals run in order, e.g. clean compile
+
+    val executorService = Executors.newFixedThreadPool(threadCount)
+    evalLog.debug(s"Created executor: ${executorService}")
+    val completionService = new ExecutorCompletionService[FutureResult](executorService)
+
+    // The scheduled and not yet finished futures (Java!)
+    var futures = List[(java.util.concurrent.Future[FutureResult], TerminalGroup)]()
+
+    try {
+
+      val interGroupDeps: Map[TerminalGroup, Seq[TerminalGroup]] = findInterGroupDeps(sortedGroups)
```

Comment:
Does the call to `findInterGroupDeps` need to be inside the `try`/`catch` block? Seems to me like this should never fail, short of a bug in Mill itself

Reply:
No need for this, also refactored away.